### DISCUSSION TITLE
Result rows are correct length, however are duplicated for each resulting row

### DISCUSSION THREAD

**Discussion by ToniTang** [sqlalchemy/sqlalchemy#9972] [Usage Questions] [answered]
Hello,

After much trial and error and finally hitting a wall, I am seeking assistance on a simple yet frustrating returned result list

Currently running `SQLAlchemy-2.0.10` and `Python 3.10.9`

I have a function that pulls ban entries from my postgresql database:
```py
async def get_ban_entries(session, steam_id: int) -> list[Row[ACBanList]] | None:
    """Retrieve a ban entry by Steam64 ID"""
    result = await session.execute(
        select(ACBanList, ACPlayerIndex.name, ACPlayerIndex.aliases, ACPlayerIndex.ban_count)
        .join(ACPlayerIndex, ACBanList.steamid == ACPlayerIndex.steamid)
        .where(ACBanList.steamid == steam_id)
        .order_by(ACBanList.date.desc()))

    # result = await session.execute(text("SELECT ac_ban_list.steamid, ac_ban_list.channel_id, "
    #                                     "ac_ban_list.ban_message_id, ac_ban_list.reason, ac_ban_list.by, "
    #                                     "ac_ban_list.server, ac_ban_list.date, ac_player_index.name, "
    #                                     "ac_player_index.aliases, ac_player_index.ban_count FROM ac_ban_list "
    #                                     "JOIN ac_player_index ON ac_ban_list.steamid = ac_player_index.steamid "
    #                                     "WHERE ac_player_index.steamid = 76561198012766295 "
    #                                     "ORDER BY ac_ban_list.date DESC"))
    data = result.all()
    for d in data:
        print(d)
    return data if len(data) > 0 else None
```
and these are my declarations for each table:
```py
class ACPlayerIndex(Base):
    __tablename__ = "ac_player_index"

    steamid: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    name: Mapped[str]
    ban_count: Mapped[int]
    aliases = Column(ARRAY(String))

    def __repr__(self) -> str:
        return f"ACPlayerIndex(steamid={self.steamid!r}, name={self.name!r}, ban_count={self.ban_count!r}, aliases={self.aliases!r}"


class ACBanList(Base):
    __tablename__ = "ac_ban_list"

    steamid: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    channel_id: Mapped[int] = mapped_column(BigInteger)
    ban_message_id: Mapped[int] = mapped_column(BigInteger)
    reason: Mapped[str]
    by: Mapped[str] = mapped_column(String(32))
    server: Mapped[str] = mapped_column(String(32))
    date: Mapped[datetime.datetime] = mapped_column(DateTime(timezone=True))

    def __repr__(self) -> str:
        return f"ACBanList(steamid={self.steamid!r}, channel_id={self.channel_id!r}, " \
               f"ban_message_id={self.ban_message_id!r}, reason={self.reason!r}, by={self.by!r}, " \
               f"server={self.server!r}, date={self.date!r})"
```
My expected output (minus the extra 3 columns from `ACPlayerIndex`: `ACPlayerIndex.name`, `ACPlayerIndex.aliases`, `ACPlayerIndex.ban_count` as I am directly fetching results from the `ac_ban_list` table)
![image](https://github.com/sqlalchemy/sqlalchemy/assets/5703654/b66603b1-a02b-45bb-b578-798cca255290)
When running the commented-out raw query (outputted/built by SQLAlchemy) included above in my first function, I get the 3 expected rows for that specific ID.

However when running it through the functions in my first code example, I get 3 entries returned, however they are all duplicated results of one row:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/5703654/35629475-f0e0-48b6-ac42-725e3c771170)

Help would be greatly appreciated. Thanks in advance

**Comment by zzzeek** [comment, marked as answer]
since this isnt a demonstration I can run, I have to think and guess what might be the problem.  I would guess that the primary key in your acbanlist table is not actually just the "steamid" column and is actually something like "(steamid, ban_message_id)".  The ORM tracks a single object per primary key value.   in both of your query outputs, "steamid" is identical in all three rows (I assume) but "ban_message_id" changes.  This means "steamid" is not the sole column in the primary key.

all guessing though of course since I dont have enough information to run the program on this end and duplicate the result.

**Reply by ToniTang** [reply]
@zzzeek just wanted to say thank you and I really appreciate your help pointing this out.

After going over both my table and declarations, I realized I had my primary key incorrectly defined for my `ac_ban_list` table.
Apologies for my overlook. I made a mistake when I was setting up the PK/FK for my table and in my declarations.

It is working as expected now :)

**Reply by zzzeek** [reply]
great, was just not very easy to read screenshots
